vpr/src/route/rr_graph_storage.cpp

```diff
@@ -709,3 +709,68 @@ t_rr_graph_view t_rr_graph_storage::view() const {
         vtr::make_const_array_view_id(edge_dest_node_),
         vtr::make_const_array_view_id(edge_switch_));
 }
+
+void t_rr_graph_storage::reorder(const vtr::vector<RRNodeId, RRNodeId>& order,
+                                 const vtr::vector<RRNodeId, RRNodeId>& inverse_order) {
+    {
+        auto old_node_storage = node_storage_;
+
+        // Reorder nodes
+        for (size_t i = 0; i < node_storage_.size(); i++) {
+            auto n = RRNodeId(i);
+            VTR_ASSERT(n == inverse_order[order[n]]);
+            node_storage_[order[n]] = old_node_storage[n];
+        }
+    }
+    {
+        auto old_node_first_edge = node_first_edge_;
+        auto old_edge_src_node = edge_src_node_;
+        auto old_edge_dest_node = edge_dest_node_;
+        auto old_edge_switch = edge_switch_;
+        RREdgeId cur_edge(0);
+
+        // Reorder edges by source node
+        for (size_t i = 0; i < node_storage_.size(); i++) {
+            node_first_edge_[RRNodeId(i)] = cur_edge;
+            auto n = inverse_order[RRNodeId(i)];
+            for (auto e = old_node_first_edge[n];
+                 e < old_node_first_edge[RRNodeId(size_t(n) + 1)];
+                 e = RREdgeId(size_t(e) + 1)) {
+                edge_src_node_[cur_edge] = order[old_edge_src_node[e]]; // == n?
+                edge_dest_node_[cur_edge] = order[old_edge_dest_node[e]];
+                edge_switch_[cur_edge] = old_edge_switch[e];
+                cur_edge = RREdgeId(size_t(cur_edge) + 1);
+            }
+        }
+    }
+    {
+        auto old_node_ptc = node_ptc_;
+        for (size_t i = 0; i < node_ptc_.size(); i++) {
+            node_ptc_[order[RRNodeId(i)]] = old_node_ptc[RRNodeId(i)];
+        }
+    }
+    {
+        auto old_node_fan_in = node_fan_in_;
+        for (size_t i = 0; i < node_fan_in_.size(); i++) {
+            node_fan_in_[order[RRNodeId(i)]] = old_node_fan_in[RRNodeId(i)];
+        }
+    }
+
+    // Check that edges are still partitioned
```

Comment:
This entire method **only** works as written if partitioning is completed.  It could be written in another fashion, however it is not.

Reply:
I'm not sure I understand this statement. The check is to make sure reordering preserves partitioning, but it does not depend on the edges to be partitioned. It preserves the same relative edge ordering for each source node e.g. edges with the same source node are moved as a block.

Reply:
It's been a while since I wrote this comment, so I'm not totally confident what I meant either.  However what I believe what I was thinking is that this entire method requires that it already be partitioned_, so really, it should be `VTR_ASSERT(partioned_)` at the top of the method, because it will explode horribly that if that isn't true.

Reply:
Why would it explode? It doesn't care. The only reason the check is there is to make sure it's safe to not re-partition.

Reply:
It's internal state variable.  If RrGraphStorage has not been partitioned, then some of the data you are working on won't be populated, and so your code will run off array boundries, etc.

Reply:
Is this documented somewhere? The reordering algorithm only requires the edge data to be correct. It doesn't care if it shuffles garbage otherwise.

Do you mean that `rr_graph_storage` is unsafe to use without calling `validate()` first?

Reply:
`rr_graph_storage` is fine to use without first partitioning the edges, in the sense that a failure to conform to the interface will result in an assertion violation.  However you are adding a method, extending the interface, and it possible for there to be a call order where you add nodes and edge data, and then reorder where the assumptions **you** have made about the state of the object are wrong.

Reply:
I've added `validate()` assertions at the beginning and end of `reorder()`.